services/oauth/v2/utils.go

```diff
@@ -0,0 +1,157 @@
+
+func GetOAuthActionStatName(stat string) string {
+	return fmt.Sprintf("oauth_action_%v", stat)
+}
+
+func (s *OAuthStats) SendTimerStats(startTime time.Time) {
+	statsTags := stats.Tags{
+		"id":              s.id,
+		"workspaceId":     s.workspaceId,
+		"rudderCategory":  s.rudderCategory,
+		"isCallToCpApi":   strconv.FormatBool(s.isCallToCpApi),
+		"authErrCategory": s.authErrCategory,
+		"destType":        s.destDefName,
+		"flowType":        string(s.flowType),
+		"action":          s.action,
+		"oauthVersion":    "v2",
+	}
+	stats.Default.NewTaggedStat(s.statName, stats.TimerType, statsTags).SendTiming(time.Since(startTime))
+}
+
+// Send count type stats related to OAuth(Destination)
+func (s *OAuthStats) SendCountStat() {
+	statsTags := stats.Tags{
+		"oauthVersion":    "v2",
+		"id":              s.id,
+		"workspaceId":     s.workspaceId,
+		"rudderCategory":  s.rudderCategory,
+		"errorMessage":    s.errorMessage,
+		"isCallToCpApi":   strconv.FormatBool(s.isCallToCpApi),
+		"authErrCategory": s.authErrCategory,
+		"destType":        s.destDefName,
+		"isTokenFetch":    strconv.FormatBool(s.isTokenFetch),
+		"flowType":        string(s.flowType),
+		"action":          s.action,
+	}
+	stats.Default.NewTaggedStat(s.statName, stats.CountType, statsTags).Increment()
+}
+
+// {input: [{}]}
+// {input: [{}, {}, {}, {}]}
+// {input: [{}, {}, {}, {}]} -> {output: [{200}, {200}, {401,authErr}, {401,authErr}]}
+func GetAuthErrorCategoryFromTransformResponse(respData []byte) (string, error) {
+	var transformedJobs []TransformerResponse
+	err := jsonfast.Unmarshal([]byte(gjson.GetBytes(respData, "output").Raw), &transformedJobs)
+	if err != nil {
+		return "", err
+	}
+	tfJob, found := lo.Find(transformedJobs, func(item TransformerResponse) bool {
+		return IsValidAuthErrorCategory(item.AuthErrorCategory)
+	})
+	if !found {
+		// can be a valid scenario
+		return "", nil
+	}
+	return tfJob.AuthErrorCategory, nil
+}
+
+func GetAuthErrorCategoryFromTransformProxyResponse(respData []byte) (string, error) {
+	transformedJobs := TransformerResponse{}
+	err := jsonfast.Unmarshal([]byte(gjson.GetBytes(respData, "output").Raw), &transformedJobs)
+	if err != nil {
+		return "", err
+	}
+	return transformedJobs.AuthErrorCategory, nil
+}
+
+func checkIfTokenExpired(secret AccountSecret, oldSecret json.RawMessage, expiryTimeDiff time.Duration, stats *OAuthStats) bool {
+	expirationDate := expirationDate{}
+	if err := json.Unmarshal(secret.Secret, &expirationDate); err != nil {
+		stats.errorMessage = "unmarshal failed"
+		stats.statName = GetOAuthActionStatName("proActive_token_refresh")
+		stats.SendCountStat()
+		return false
+	}
+	if expirationDate.ExpirationDate != "" && isTokenExpired(expirationDate.ExpirationDate, expiryTimeDiff, stats) {
+		return true
+	}
+	if router_utils.IsNotEmptyString(string(oldSecret)) {
+		if bytes.Equal(secret.Secret, oldSecret) {
+			return true
+		}
+	}
```

Comment:
[optional]:  We can probably remove one level of nesting in here.
```suggestion
	if !router_utils.IsNotEmptyString(string(oldSecret)) {
		return false
	}
	if bytes.Equal(secret.Secret, oldSecret) {
		return true
	}
```